README
This project will provide a base to write tests. The aim is to stress a
filesystem. As there is many type of filesystem and that some of them are
also networked (meaning that all computing is not only done by the machine
where the program runs) there is no special data retrieved (say some cpu
utilisation, I/O bandwidth and so on). This should be done with other tools
such as munin, nagios, and so on... .

Every one is encouraged to write a test for a specific case.

When creating a new test please note the following : 
 - Add new files (my_test.c and include/my_test.h) to src/Makefile.am
 - Include my_test.h in gstressfs.h (at the end)
 - All tests should be considered as atomic (the test framework does so)
   but you can write non atomic tests, it only means that they'll run
   much more times (see creation_file as an example)
 - Keep in mind that your atomic test will be threaded, so do not write
   in your private datas (as there is no mutex capabilities and no
   memory copy, you may experiment funny things coarse to debug !).
 

ChangeLog
06.08.2009 Olivier Delhomme <olivier.delhomme@free.fr>
       * Python seem very simple and powerfull, this is getting better and
	     better

04.08.2009 Olivier Delhomme <olivier.delhomme@free.fr>
	   * beginning things in python

23.08.2009 Olivier Delhomme <olivier.delhomme@free.fr>
	   * Commited holliday's work and creating a new python
             branch and moving all python stuff to it !


27.04.2008 Olivier Delhomme <olivierdelhomme@gmail.com>
	   * Modified the calls to the libduptools to
	     refer the new 0.0.4 API

06.03.2008 Olivier Delhomme <olivierdelhomme@gmail.com>
	   * Added tests to avoid seg fault when the filesystem
	     hangs out (unable to create a new directory)
	     Added a new test in file_creation that creates one file
	     of a defined size (from 4k to 1M)
	
05.03.2008 Olivier Delhomme <olivierdelhomme@gmail.com>
	   * Added a new menu function that group functions together
	   * Added varying function list that collects functions that
	     may help to vary some parameters. Thoses functions works
	     with run_one_test_many_threaded_increasing function.
	   * In file_creation.c, for increased performances moved a 
	     g_malloc and g_free couple out of the test function in 
	     order that it only runs one time instead of hundreds of 
	     thousand ! If we want some buffer varying function we may
	     add this couple to the varying function itself !
	   * Passed file_size type from guint to guint64.
	   * rusage statistics collection (Still a lot of work to do
	     with threads and for printing them out !)
	
22.02.2008 Olivier Delhomme <olivierdelhomme@gmail.com>
	   * Simplified the exec directory name handling
	   * All tests are running in their own sub-directories
	   * Changed the initialisation for the tests in order
	     that each initialisation can initiate more than on test
	   * Clarified a bit README file in order that people understand
	     howto make new tests
	   * Changed #define name to be more clear and less confusing
	
21.02.2008 Olivier Delhomme <olivierdelhomme@gmail.com>
	   * Added a percent gfloat in test_t structure to reflect the 
	     percentage realised of the test running (until it gets an 
	     error).
	   * More thread safe
	   * Added test menus and submenus (one test, many tests) still to
	     code threaded tests, all tests and more ?
	
20.02.2008 Olivier Delhomme <olivierdelhomme@gmail.com>
	   * Now depends on libduptools and no more on liblats
	
19.02.2008 Olivier Delhomme <olivierdelhomme@gmail.com>
	   * Added private data to the file_creation test (as an example)
	
18.02.2008 Olivier Delhomme <olivierdelhomme@gmail.com>
	   * Added a timer to know the execution time
	   * Added a threaded test system
	
08.02.2008 Olivier Delhomme <olivierdelhomme@gmail.com>
	   * Added one basic test (file_creation)
	   * Added basic testing features (multiples times, incrementation)
	
07.02.2008 Olivier Delhomme <olivierdelhomme@gmail.com>
	   * Added first glade interface
	
06.02.2008 Olivier Delhomme <olivierdelhomme@gmail.com>
	   * Project creation & skeletton
	     - depends on liblats and libglade


autoclean.sh
#!/bin/bash
#
# Author : Olivier Delhomme
# Licence : GNU GPL v2
#
if [ -f Makefile ];
 then
    make distclean
fi;
rm -f aclocal.m4
echo "rm -f aclocal.m4"
rm -fr autom4te.cache
echo "rm -fr autom4te.cache"
rm -f config.guess
echo "rm -f config.guess"
rm -f config.sub
echo "rm -f config.sub"
rm -f depcomp
echo "rm -f depcomp"
rm -f INSTALL
echo "rm -f INSTALL"
rm -f install-sh
echo "rm -f install-sh"
rm -f ltmain.sh
echo "rm -f ltmain.sh"
rm -f missing
echo "rm -f missing"
rm -f src/*~ src/*#
echo "rm -f src/*~ src/*#"
rm -f *~ *#
echo "rm -f *~ *#"
rm -f Makefile.in src/Makefile.in
echo "rm -f Makefile.in src/Makefile.in"
rm -f configure
echo "rm -f configure"
rm -f config.status config.log
echo "rm -f config.status config.log"
rm -f libtool
echo "rm -f libtool"
find -name *~ -exec rm {} \;
echo "find -name *~ -exec rm {} \;"
find -name *.gcd[ao] -exec rm -f {} \;
echo "find -name *.gcd[ao] -exec rm -f {} \;"
find -name *.gcov -exec rm -f {} \;
echo "find -name *.gcov -exec rm -f {} \;"
find -name *.gcno -exec rm -f {} \;
echo "find -name *.gcno -exec rm -f {} \;"
rm -f intltool-extract.in
echo "rm -f intltool-extract.in"
rm -f intltool-merge.in
echo "rm -f intltool-merge.in"
rm -f intltool-update.in
echo "rm -f intltool-update.in"
rm -f intltool-extract
echo "rm -f intltool-extract"
rm -f intltool-merge
echo "rm -f intltool-merge"
rm -f intltool-update
echo "rm -f intltool-update"
rm -f src/gmon.out
echo "rm -f src/gmon.out"



src/stress.py
#!/usr/bin/env python
# -*- coding: utf8 -*-
#
#  Module that defines a stress class
#
#  (C) Copyright 2009 Olivier Delhomme
#  e-mail : olivier.delhomme@free.fr
# 
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2, or (at your option)
#  any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software Foundation,
#  Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
#

"""
module stress
"""

import os, time

class TestSuite :
	
	name = ''
	path = ''
	testlist = []
	
	def __init__(self, name, path) :
		self.name = name
		self.path = path


	def add_test(self, test) :
		self.testlist.append(test)


	def run_test_suite(self) :
		max = len(self.testlist)
		for i in range(max) :
			test = self.testlist[i]
			test.start(self.path)

						
	def set_path(self, path) :
		self.path = path


	def print_stats(self) :
		print 'Results for TestSuite "' + self.name + '"'
		max = len(self.testlist)
		for i in range(max) :
			print
			self.testlist[i].print_stats()
			
	def list_tests(self) :
		print 'Test list for TestSuite "' + self.name + '"'
		max = len(self.testlist)
		for i in range(max) :
			print str(i).ljust(3) + ' : ' + self.testlist[i].name
		
# Class TestSuite


class Test :
	
	def __init__(self, name, function) :
		self.name = name
		self.execute = function
		self.times = []
	
	
	def default_test(self) :
		print 'This is the default test'
	
	
	def start(self, path) :
		
		current_path = os.getcwd()
		os.chdir(path)
		
		begin_cpu = time.clock()
		begin_time = time.time()
		
		self.execute(path)
		
		end_time = time.time()
		end_cpu = time.clock()
		
		os.chdir(current_path)
		
		a_time = end_cpu - begin_cpu, end_time - begin_time
		self.times.append(a_time)


	def print_stats(self) :
		
		nb_tests = len(self.times)
		avg_cpu = 0
		avg_real = 0
		print 'Results for test "' + self.name + '"'
		print 'Tests'.ljust(8) + ' . ' + 'CPU   '.rjust(8) + '.' + 'real time'.rjust(42)
		for i in range(nb_tests) :
			cpu_time, real_time = self.times[i]
			avg_cpu += cpu_time
			avg_real += real_time
			print str(i).ljust(8) + ' : ' + str(cpu_time).rjust(8) + ' ' + str(real_time).rjust(42)
		
		print 'Averages : ' + str(avg_cpu/nb_tests).rjust(8) + ' ' + str(avg_real/nb_tests).rjust(42)


	name = ''		     	# name for the test
	execute = default_test	# function being executed by the test
	times = []              # execution times tuples (cpu, real_time)
	
	
# Class Test



src/stressfs.py
#!/usr/bin/env python
# -*- coding: utf8 -*-
#
#  Tools to stress your filesystem. Use at your own risks !
#
#  (C) Copyright 2009 Olivier Delhomme
#  e-mail : olivier.delhomme@free.fr
# 
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2, or (at your option)
#  any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software Foundation,
#  Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
#
import os, sys, getopt, stress


# Help message for main program
def usage() :
  print """
  NAME
      stressfs
  
  SYNOPSIS
      stressfs [OPTION]... 
      
  DESCRIPTION
      
      
  OPTION
      -h, --help
        Show this help
        
  EXAMPLE

  """
#


def parse_command_line(stressfs) :
	short_options = "hl"
	long_options = ["help", "list"]

	# Read options and arguments  
	try:
		opts, args = getopt.getopt(sys.argv[1:], short_options, long_options)
	except getopt.GetoptError, err:
		# print help information and exit:
		print str(err) # will print something like "option -a not recognized"
		usage()
		sys.exit(2)
	
	for opt, arg in opts:
#		print "opt = " + opt
#		print "arg = " + arg
		
		if opt in ("-h", "--help") :
			usage()
			sys.exit(0)
		elif opt in ("l", "--list") :
			stressfs.list_tests()
			sys.exit(0)

	stressfs.run_test_suite()
	stressfs.print_stats()
			
			
#

def my_first_test(path) :
	
	for i in range(10) :
		dir = path + '/' + str(i)
		try :
			os.mkdir(dir)
		except OSError, err :
			print str(err)
#

def my_second_test(path) :
	
	print 'A second test !!! $$ ' + path
	j = 1
	for i in range(10000) :
		j *= 2
#


def main () :

	test1 = stress.Test('directory creation', my_first_test)
	test2 = stress.Test('deuxieme', my_second_test)

	stressfs = stress.TestSuite('Files', '/home/dup/tmp/stressfs')
	stressfs.add_test(test1)
	stressfs.add_test(test2)

	parse_command_line(stressfs)
	
# main function


if __name__=="__main__" :
  main()

